app/gui/view/graph-editor/src/component/visualization/foreign/java_script/source.rs

```diff
@@ -0,0 +1,165 @@
+//! Contains [`Sources`] definition, a set of JS source code files with attached source map.
+use crate::prelude::*;
+
+use crate::visualization::Project;
+
+use sourcemap::SourceMapBuilder;
+use std::path::PathBuf;
+
+
+
+// ============================
+// === Visualizations paths ===
+// ============================
+
+/// Path prefixes for source files depending on `Project` enum.
+/// These prefixes form a path to the source file displayed in the DevTools. We provide a specific
+/// set of prefixes to make sources navigation in DevTools easier. As we have access to relative
+/// file paths only, it would be hard to understand what relative path corresponds to otherwise.
+mod path_prefix {
+    use super::*;
+
+    const BUILTIN: &str = "visualization/builtin";
+    const CURRENT_PROJECT: &str = "visualization/project";
+    const LIBRARY: &str = "visualization/library";
+
+    /// Turn `Project` enum into a path prefix.
+    pub fn from_project(project: &Project) -> PathBuf {
+        match *project {
+            Project::Builtin => PathBuf::from(BUILTIN),
+            Project::CurrentProject => PathBuf::from(CURRENT_PROJECT),
+            Project::Library(ref name) => {
+                let name: &str = name.as_ref();
+                PathBuf::from(LIBRARY).join(name)
+            }
+        }
+    }
+}
+
+
+
+// ===============
+// === Sources ===
+// ===============
+
+/// Internal identifier of the source file.
+type SourceId = u32;
+
+/// A set of JS source code files with attached
+/// [source map](https://blog.teamtreehouse.com/introduction-source-maps).
+///
+/// It is used to concatenate several source files into a single one and preserve
+/// the information about original file names and line numbers.
+pub struct Sources {
+    code:               String,
+    lines_count:        u32,
+    source_map_builder: SourceMapBuilder,
+    files:              Vec<(PathBuf, SourceId)>,
+}
+
+impl Debug for Sources {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Sources")
+            .field("code", &self.code)
+            .field("lines_count", &self.lines_count)
+            .field("files", &self.files)
+            .finish()
+    }
+}
+
+impl Default for Sources {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Sources {
+    /// Constructor. Typically [`Sources::from_files`] should be used instead.
+    pub fn new() -> Self {
```

Comment:
Would `empty` be more descriptive as a constructor name here instead of `new`?

Reply:
Yeah, I think so. Thank you, renamed